File: mechanical/src/main/scala/resonantinduction/mechanical/gear/PartGear.java
package resonantinduction.mechanical.gear;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.MovingObjectPosition;
import net.minecraft.world.World;
import net.minecraftforge.common.ForgeDirection;
import resonant.api.grid.INode;
import resonant.api.grid.INodeProvider;
import resonant.lib.multiblock.IMultiBlockStructure;
import resonant.lib.multiblock.MultiBlockHandler;
import resonant.lib.utility.WrenchUtility;
import resonantinduction.core.Reference;
import resonantinduction.core.interfaces.IMechanicalNode;
import resonantinduction.core.resource.ItemHandCrank;
import resonantinduction.mechanical.Mechanical;
import resonantinduction.mechanical.energy.grid.MechanicalNode;
import resonantinduction.mechanical.energy.grid.PartMechanical;
import codechicken.lib.vec.Cuboid6;
import codechicken.lib.vec.Rotation;
import codechicken.lib.vec.Transformation;
import codechicken.lib.vec.Vector3;
import codechicken.microblock.FaceMicroClass;
import codechicken.multipart.ControlKeyModifer;
import codechicken.multipart.TMultiPart;
import codechicken.multipart.TileMultipart;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;

/** We assume all the force acting on the gear is 90 degrees.
 * 
 * @author Calclavia */
public class PartGear extends PartMechanical implements IMultiBlockStructure<PartGear>
{
    public static Cuboid6[][] oBoxes = new Cuboid6[6][2];

    static
    {
        oBoxes[0][0] = new Cuboid6(1 / 8D, 0, 0, 7 / 8D, 1 / 8D, 1);
        oBoxes[0][1] = new Cuboid6(0, 0, 1 / 8D, 1, 1 / 8D, 7 / 8D);
        for (int s = 1; s < 6; s++)
        {
            Transformation t = Rotation.sideRotations[s].at(Vector3.center);
            oBoxes[s][0] = oBoxes[0][0].copy().apply(t);
            oBoxes[s][1] = oBoxes[0][1].copy().apply(t);
        }
    }

    private boolean isClockwiseCrank = true;
    private int manualCrankTime = 0;
    private int multiBlockRadius = 1;

    public PartGear()
    {
        super();
        node = new GearNode(this);
       
    }

    @Override
    public void update()
    {
        super.update();

        if (!this.world().isRemote)
        {
            if (manualCrankTime > 0)
            {
                node.apply(this, isClockwiseCrank ? 15 : -15, isClockwiseCrank ? 0.025f : -0.025f);
                manualCrankTime--;
            }
        }
        getMultiBlock().update();
    }

    @Override
    public void checkClientUpdate()
    {
        if (getMultiBlock().isPrimary())
            super.checkClientUpdate();
    }

    @Override
    public boolean activate(EntityPlayer player, MovingObjectPosition hit, ItemStack itemStack)
    {
        if (itemStack != null && itemStack.getItem() instanceof ItemHandCrank)
        {
            if (!world().isRemote && ControlKeyModifer.isControlDown(player))
            {
                getMultiBlock().get().node.torque = -getMultiBlock().get().node.torque;
                getMultiBlock().get().node.angularVelocity = -getMultiBlock().get().node.angularVelocity;
                return true;
            }

            isClockwiseCrank = player.isSneaking();
            getMultiBlock().get().manualCrankTime = 20;
            world().playSoundEffect(x() + 0.5, y() + 0.5, z() + 0.5, Reference.PREFIX + "gearCrank", 0.5f, 0.9f + world().rand.nextFloat() * 0.2f);
            player.addExhaustion(0.01f);
            return true;
        }

        if (WrenchUtility.isWrench(itemStack))
        {
            getMultiBlock().toggleConstruct();
            return true;
        }

        return super.activate(player, hit, itemStack);
    }

    @Override
    public void preRemove()
    {
        super.preRemove();
        getMultiBlock().deconstruct();
    }

    /** Is this gear block the one in the center-edge of the multiblock that can interact with other
     * gears?
     * 
     * @return */
    public boolean isCenterMultiBlock()
    {
        if (!getMultiBlock().isConstructed())
        {
            return true;
        }

        universalelectricity.api.vector.Vector3 primaryPos = getMultiBlock().getPrimary().getPosition();

        if (primaryPos.intX() == x() && placementSide.offsetX == 0)
        {
            return true;
        }

        if (primaryPos.intY() == y() && placementSide.offsetY == 0)
        {
            return true;
        }

        if (primaryPos.intZ() == z() && placementSide.offsetZ == 0)
        {
            return true;
        }

        return false;
    }

    @Override
    protected ItemStack getItem()
    {
        return new ItemStack(Mechanical.itemGear, 1, tier);
    }

    @Override
    @SideOnly(Side.CLIENT)
    public void renderDynamic(Vector3 pos, float frame, int pass)
    {
        if (pass == 0)
        {
            RenderGear.INSTANCE.renderDynamic(this, pos.x, pos.y, pos.z, tier);
        }
    }

    @Override
    public String getType()
    {
        return "resonant_induction_gear";
    }

    @Override
    public void load(NBTTagCompound nbt)
    {
        super.load(nbt);
        getMultiBlock().load(nbt);
    }

    @Override
    public void save(NBTTagCompound nbt)
    {
        super.save(nbt);
        getMultiBlock().save(nbt);
    }

    /** Multiblock */
    private GearMultiBlockHandler multiBlock;

    @Override
    public universalelectricity.api.vector.Vector3[] getMultiBlockVectors()
    {
        Set<universalelectricity.api.vector.Vector3> vectors = new HashSet<universalelectricity.api.vector.Vector3>();
        ForgeDirection dir = placementSide;

        universalelectricity.api.vector.Vector3 rotationalAxis = universalelectricity.api.vector.Vector3.UP();

        if (placementSide == ForgeDirection.NORTH || placementSide == ForgeDirection.SOUTH)
        {
            rotationalAxis = universalelectricity.api.vector.Vector3.EAST();
        }
        else if (placementSide == ForgeDirection.WEST || placementSide == ForgeDirection.EAST)
        {
            rotationalAxis = universalelectricity.api.vector.Vector3.SOUTH();
        }

        for (int x = -1; x <= 1; x++)
        {
            for (int z = -1; z <= 1; z++)
            {
                universalelectricity.api.vector.Vector3 vector = new universalelectricity.api.vector.Vector3(x, 0, z);
                vector.rotate(90, rotationalAxis);
                vector = vector.round();
                vectors.add(vector);
            }
        }

        return vectors.toArray(new universalelectricity.api.vector.Vector3[0]);
    }

    @Override
    public World getWorld()
    {
        return world();
    }

    @Override
    public void onMultiBlockChanged()
    {
        if (world() != null)
        {
            tile().notifyPartChange(this);

            if (!world().isRemote)
            {
                sendDescUpdate();
            }
        }
    }

    @Override
    public MultiBlockHandler<PartGear> getMultiBlock()
    {
        if (multiBlock == null)
            multiBlock = new GearMultiBlockHandler(this);

        return multiBlock;
    }

    @Override
    public INode getNode(Class<? extends INode> nodeType, ForgeDirection from)
    {
        if (nodeType.isAssignableFrom(node.getClass()))
            return getMultiBlock().get().node;
        return null;
    }

    /** Multipart Bounds */
    @Override
    public Iterable<Cuboid6> getOcclusionBoxes()
    {
        return Arrays.asList(oBoxes[this.placementSide.ordinal()]);
    }

    @Override
    public int getSlotMask()
    {
        return 1 << this.placementSide.ordinal();
    }

    @Override
    public Cuboid6 getBounds()
    {
        return FaceMicroClass.aBounds()[0x10 | this.placementSide.ordinal()];
    }

    @Override
    @SideOnly(Side.CLIENT)
    public Cuboid6 getRenderBounds()
    {
        return Cuboid6.full.copy().expand(multiBlockRadius);
    }

    @Override
    public String toString()
    {
        return "[PartGear]" + x() + "x " + y() + "y " + z() + "z " + getSlotMask() + "s ";
    }
}

File: src/main/scala/edx/mechanical/mech/gear/PartGear.scala
package edx.mechanical.mech.gear

import java.util

import codechicken.lib.vec.{Cuboid6, Vector3}
import codechicken.microblock.FaceMicroClass
import cpw.mods.fml.relauncher.{Side, SideOnly}
import edx.core.Reference
import edx.core.prefab.part.CuboidShapes
import edx.mechanical.MechanicalContent
import edx.mechanical.mech.PartMechanical
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.MovingObjectPosition
import net.minecraft.world.World
import net.minecraftforge.common.util.ForgeDirection
import resonantengine.api.tile.node.INode
import resonantengine.lib.prefab.tile.multiblock.reference.IMultiBlockStructure
import resonantengine.lib.utility.WrenchUtility

/**
 * We assume all the force acting on the gear is 90 degrees.
 *
 * @author Calclavia
 */
class PartGear extends PartMechanical with IMultiBlockStructure[PartGear]
{
  val multiBlock = new GearMultiBlockHandler(this)
  var multiBlockRadius = 1

  //Constructor
  mechanicalNode = new NodeGear(this)

  mechanicalNode.onVelocityChanged = () =>
  {
    if (getMultiBlock.isPrimary)
      if (world != null) sendPacket(1)

    if (mechanicalNode.angularVelocity == 0)
      if (world != null) sendPacket(2)
  }

  mechanicalNode.onGridReconstruct = () =>
  {
    if (world != null)
    {
      sendPacket(1)
      sendPacket(2)
    }
  }

  //TODO: Can we not have update ticks here?
  override def update()
  {
    super.update()
    getMultiBlock.update()
  }

  override def activate(player: EntityPlayer, hit: MovingObjectPosition, itemStack: ItemStack): Boolean =
  {
    if (itemStack != null && itemStack.getItem.isInstanceOf[ItemHandCrank])
    {
      mechanicalNode.accelerate((if (player.isSneaking) 1 else -1) * 1000)
      world.playSoundEffect(x + 0.5, y + 0.5, z + 0.5, Reference.prefix + "gearCrank", 0.5f, 0.9f + world.rand.nextFloat * 0.2f)
      player.addExhaustion(0.02f)
      return true
    }

    if (WrenchUtility.isWrench(itemStack))
    {
      getMultiBlock.toggleConstruct()
      return true
    }

    return super.activate(player, hit, itemStack)
  }

  override def preRemove()
  {
    super.preRemove()
    getMultiBlock.deconstruct()
  }

  /**
   * Is this gear block the one in the center-edge of the multiblock that can interact with other gears?
   * @return Returning true implies that this gear is able to connect to other ones side-by-side.
   */
  def isCenterMultiBlock: Boolean =
  {
    if (!getMultiBlock.isConstructed)
    {
      return true
    }
    val primaryPos = getMultiBlock.getPrimary.getPosition
    return (primaryPos.xi == x && placementSide.offsetX == 0) || (primaryPos.yi == y && placementSide.offsetY == 0) || (primaryPos.zi == z && placementSide.offsetZ == 0)
  }

  @SideOnly(Side.CLIENT)
  override def renderDynamic(pos: Vector3, frame: Float, pass: Int)
  {
    if (pass == 0)
      RenderGear.renderDynamic(this, pos.x, pos.y, pos.z, tier)
  }

  override def load(nbt: NBTTagCompound)
  {
    super.load(nbt)
    getMultiBlock.load(nbt)
  }

  override def save(nbt: NBTTagCompound)
  {
    super.save(nbt)
    getMultiBlock.save(nbt)
  }

  override def getMultiBlock: GearMultiBlockHandler = multiBlock

  override def getMultiBlockVectors: java.util.List[resonantengine.lib.transform.vector.Vector3] = new resonantengine.lib.transform.vector.Vector3().getAround(this.world, placementSide, 1)

  def getWorld: World =
  {
    return world
  }

  def onMultiBlockChanged()
  {
    if (world != null)
    {
      tile.notifyPartChange(this)

      if (!world.isRemote)
      {
        mechanicalNode.reconstruct()
        sendDescUpdate()
      }
    }
  }

  override def getNode[N <: INode](nodeType: Class[_ <: N], from: ForgeDirection): N =
  {
    if (nodeType.isAssignableFrom(mechanicalNode.getClass) && from == placementSide)
      return getMultiBlock.get.mechanicalNode.asInstanceOf[N]

    return null.asInstanceOf[N]
  }

  /** Multipart Bounds */
  def getOcclusionBoxes: java.lang.Iterable[Cuboid6] =
  {
    val list: java.util.List[Cuboid6] = new util.ArrayList[Cuboid6];
    for (v <- CuboidShapes.panel(this.placementSide.ordinal))
    {
      list.add(v)
    }
    return list
  }

  def getBounds: Cuboid6 =
  {
    return FaceMicroClass.aBounds(0x10 | this.placementSide.ordinal)
  }

  @SideOnly(Side.CLIENT)
  override def getRenderBounds: Cuboid6 = Cuboid6.full.copy.expand(multiBlockRadius)

  override def toString = "[PartGear]" + x + "x " + y + "y " + z + "z " + getSlotMask + "s "

  def getSlotMask: Int =
  {
    return 1 << this.placementSide.ordinal
  }

  protected def getItem: ItemStack =
  {
    return new ItemStack(MechanicalContent.itemGear, 1, tier)
  }
}